hardware/VirtualThermostat.cpp

```diff
@@ -0,0 +1,715 @@
+			break;
+		sec_counter++;
+
+		if (sec_counter % 12 == 0) {
+			m_LastHeartbeat = mytime(NULL);
+		}
+		time_t atime = mytime(NULL);
+		struct tm ltime;
+		localtime_r(&atime, &ltime);
+#if 1
+		if (ltime.tm_min != m_ScheduleLastMinute)
+		{
+			ScheduleThermostat(ltime.tm_min);
+			m_ScheduleLastMinute = ltime.tm_min;
+		}
+#else
+		//test : call every 6 secconds
+		if (ltime.tm_sec % 6 == 0)
+		{
+			ScheduleThermostat(ltime.tm_sec / 6);
+			m_ScheduleLastMinute = ltime.tm_sec;
+		}
+#endif
+	}
+	Log(LOG_STATUS, "VirtualThermostat: Worker stopped...");
+}
+bool VirtualThermostat::WriteToHardware(const char* pdata, const unsigned char length)
+{
+	return true;
+}
+
+//time for the power time modulation in minute
+#define MODULATION_DURATION 10
+//  number of step in percent %
+#define MODULATION_STEP     10
+
+//guve the output power swicth value in the time from the modulation percent
+const int ThermostatOutput[MODULATION_STEP + 1][MODULATION_DURATION] = {
+	//            0 1 2 3 4 5 6 7 8 9    
+	/* 000 % */ { 0,0,0,0,0,0,0,0,0,0 },
+	/* 010 % */ { 1,0,0,0,0,0,0,0,0,0 },
+	/* 020 % */ { 1,1,0,0,0,0,0,0,0,0 },
+	/* 030 % */ { 1,1,1,0,0,0,0,0,0,0 },
+	/* 040 % */ { 1,1,1,1,0,0,0,0,0,0 },
+	/* 050 % */ { 1,1,1,1,1,0,0,0,0,0 },
+	/* 060 % */ { 1,1,1,1,1,1,0,0,0,0 },
+	/* 070 % */ { 1,1,1,1,1,1,1,0,0,0 },
+	/* 080 % */ { 1,1,1,1,1,1,1,1,0,0 },
+	/* 090 % */ { 1,1,1,1,1,1,1,1,1,0 },
+	/* 100 % */ { 1,1,1,1,1,1,1,1,1,1 },
+
+};
+
+//compute the thermostat output switch value 
+// the output is modulated in a time period of 10 minute
+// each minute , the output is activated dependeing time and percent.
+//input : Min : the minute counter 0..59
+//input : PowerPercent : the power modulation  0..100%
+//output 1 : activated
+
+int VirtualThermostat::ComputeThermostatOutput(int Min, int PowerPercent)
+{
+	if (PowerPercent > 100) PowerPercent = 100;
+	//get row value for ThermostatOutput
+	PowerPercent = (PowerPercent) / MODULATION_STEP;
+	int switchValue = ThermostatOutput[PowerPercent][Min % MODULATION_DURATION];
+	return switchValue;
+}
+
+//return the poqer modulation in function of Room , Exterior and Target Temperature,
+int VirtualThermostat::ComputeThermostatPower(int index, double RoomTemp, double TargetTemp, double CoefProportional, double CoefIntegral)
+{
+	int PowerModulation = 0;
+	double DeltaTemp = TargetTemp - RoomTemp;
+	CircularBuffer* Delta = DeltaTemps[index];
+	if (Delta == 0)
+	{
+		Delta = new CircularBuffer(INTEGRAL_DURATION);
+		DeltaTemps[index] = Delta;
+	}
+	//put last value
+	Delta->Put(DeltaTemp);
+	//need heating
+	if (DeltaTemp > 0)
+	{
+		//PID regulation
+		// coef integral is the sum of delta temperature since last 10 minutes.
+		PowerModulation = int(DeltaTemp * CoefProportional + Delta->GetSum() * CoefIntegral / INTEGRAL_DURATION);
+	}
+	else
+		Delta->Clear(); //clear integral part
+
+	if (PowerModulation > 100) PowerModulation = 100;
+	if (PowerModulation < 0) PowerModulation = 0;
+	return PowerModulation;
```

Comment:
https://en.cppreference.com/w/cpp/algorithm/clamp

Reply:
ok ! done . i dont know

Reply:
std::clamp  not compile on raspbery pi : c++17 !